### ISSUE TITLE
fix - dust sensor which doesn't work all the time

### ISSUE THREAD

**Issue by RobertLucian** [DexterInd/GrovePi#408]
There's already a PR meant for the dust sensor, but it doesn't address all problems: https://github.com/DexterInd/GrovePi/issues/365

There's also a post on the forum that addresses the issues with the dust sensor:
https://forum.dexterindustries.com/t/dust-sensor-in-python/1767/19

Finally, there is an improvement that was asked on this topic on the forum:
https://forum.dexterindustries.com/t/improving-dust-sensor-resolution-from-30-seconds-to-10-seconds/4912/11

**Comment by RobertLucian** [comment]
I have noticed that the timing in the `grovepi.py` script isn't handled in all places. 

I realized that when I tried to poll the dust sensor faster at higher rates without adding any delay between the I2C calls. When I would do that, the dust sensor function would no longer notify me of new values through the `databack[0]` parameter:
```
def dustSensorRead():
	"""
	By default, the sample rate is set to 1 at every 30 seconds and this
	function was written only for that interval.

	If you wish to use a different
	interval, then use dustSensorReadMore function. To set a
	different interval, use setDustSensrInterval function.
	"""
	write_i2c_block(address, dus_sensor_read_cmd + [unused, unused, unused])
	time.sleep(.2)
	#read_i2c_byte(address)
	#number = read_i2c_block(address)
	#return (number[1] * 256 + number[2])
	data_back= bus.read_i2c_block_data(address, 1)[0:6]
	#print data_back[:4]
	if data_back[0]!=255:
		lowpulseoccupancy=(data_back[3] * 65536 + data_back[2] * 256 + data_back[1])
		#print [data_back[0],lowpulseoccupancy]
		return [data_back[0], lowpulseoccupancy]
	else:
		return [-1,-1]
```

Looking at this code of  the `grovepi.py`, I noticed that there's no `time.sleep` after the second I2C call. After adding one in there the dust sensor started working really well. So the problem is that the for loop in the firmware needs some time to run it and when 2 consecutive I2C calls are made with no delay between them, things will get messed. 

What's interesting is that a bunch of functions don't have the 2nd delay added for the 2nd call, which might mean this is the cause for #.

**Comment by RobertLucian** [comment]
Seems like with the new changes that allow a much faster polling rate, the dust sensor no longer reports the correct values when the polling rate is slightly higher than what it used to be. Since this issue ought not to be fixed by adding delays, I resorted to finding the bug and seeing where it actually gets stuck.

After digging some time into this issue, I came to the conclusion that the problem comes from the variable that holds the low pulse occupancy value, whose value gets changed w/o having to when it's being used in the main loop and in the ISR. The problem doesn't come from not being a volatile variable or anything humanly imaginable, but it comes from the fact it was using `unsigned long` as a data type. It seems like when `unsigned longs`  (volatile) are used in both ISRs and the main loop, the variable change's its value without any sense. 

Changing the data type to a 16 bit one seems to mitigate the issue so most likely this is a compiler issue.
